### PR TITLE
203 view liked posts & 198 view liked posts on profile

### LINKED ISSUE #198: [198] View Liked Posts On Profile
**Current state**
There is no back-end functionality for fetching all posts a user liked, and this isn't shown on the profile page.

**Expected state**
Users can view their liked posts on their profiles.

**Todo**
Backend
- [ ] Fetching liked posts of a user

Frontend
- [ ] Showing liked posts in frontend
- [ ] Show amount of likes for these posts

**Necessities**
n/a

**Branch**
198-view-liked-posts

### LINKED ISSUE #203: [203] View Liked Posts
Closes #XXX

**Current state**
It is not possible to view who liked the current post.

**Expected state**
Visitors can view the users who liked the post they're reading.

**TO DO**
- [x] Back-end functionality for fetching all likes
- [x] Basic front-end styling for 
- [x] Front-end functionality for showing likes with links to their profiles

**Necessities**
n/a

**Branch**
198-view-liked-posts

### PR DESCRIPTION
**Pull request by jerohero**
closes #203
closed #198

 **What is it supposed to do**
Everyone can see who liked a post and can see what posts a user liked on their profile (yes I put two different use cases in one branch)

 **How can we test your branch**
- Go to a post and click on the amount of a like to open the list (shouldn't be possible if there are no likes)
- Click on a profile in the list, test whether it's sorting properly or whatever
- Go to a profile and look at the liked posts. See if this updates properly and is sorted.

 **What broweser have been checked**
- [x] Chrome
- [x] Firefox
- [ ] Opera
- [ ] Edge
- [ ] Safari

### CODE REVIEW

**Review by AjobK — CHANGES_REQUESTED** [review]
Change to fontawesome icons, see screenshot for style reference too:
image

Rename feathers to likes

Make overlay fixed, not absolute

**Review by AjobK — CHANGES_REQUESTED** [review]
image

When hallway testing it appeared that spamming the like button results in multiple likes by a single person.

**Comment by jerohero** [comment]
> image
> 
> When hallway testing it appeared that spamming the like button results in multiple likes by a single person.

I was unable to recreate this issue by spamming the like button, and forcing a like request from DevTools while the post is liked wouldn't succeed either.

chrome_Ajk6rjYwMi

**Review by AjobK — CHANGES_REQUESTED** [review]
image
Unable to scroll in this window

Please make the profile_id and post_id combination unique in the post_like table, so there will never be a duplicate like (In other words, one person liking twice).

Also turn the 'X' on the right into an icon.
image
I suggest this one: https://fontawesome.com/icons/times-circle?style=solid

Change the view into the example previously shown in discord, except for the 'Feathers' part which should be 'Likes' and obviously the icons:
image

File: backend/app/dao/postDao.ts [hunk 1]
```diff
@@ -56,6 +56,17 @@ class PostDAO {
         return await repository.find({ where: {post: id}, relations: ['post', 'profile'] })
     }
 
+    public async getRecentUserLikesByProfileId(profileId: number, limit): Promise<any> {
+        if (!profileId)
+            return null
```
Comment (AjobK) [inline backend/app/dao/postDao.ts]: Newline under return

File: backend/app/dao/postDao.ts [hunk 2]
```diff
@@ -56,6 +56,17 @@ class PostDAO {
...
+        const repository = await DatabaseConnector.getRepository('PostLike')
+        const likes = await repository.find({ where: {profile: profileId}, relations: ['post', 'profile'], take: limit})
```
Comment (AjobK) [inline backend/app/dao/postDao.ts]: newline under declaration

File: backend/app/dao/postDao.ts [hunk 3]
```diff
@@ -56,6 +56,17 @@ class PostDAO {
...
+        if (!likes || likes.length <= 0) return
```
Comment (AjobK) [inline backend/app/dao/postDao.ts]: newline under

File: backend/app/dao/postDao.ts [hunk 4]
```diff
@@ -56,6 +56,17 @@ class PostDAO {
...
+        return likes.sort((a, b) => {
```
Comment (AjobK) [inline backend/app/dao/postDao.ts]: a and b is cryptic. Please use understandable variable names

File: backend/app/service/postService.ts [hunk 5]
```diff
@@ -188,11 +183,27 @@ class PostService {
...
+
+        if (req.params.username && foundLikes) {
+            const likes = []
+            foundLikes.forEach((like) => {
+                likes.push(like.post)
+            })
+
+            return res.status(200).json(likes)
+        }
+        else
```
Comment (AjobK) [inline backend/app/service/postService.ts]: Please use either shorthand or the extended version. Don't combine if () {} with a normal else without curly braces. In this case use curly braces for the else

File: backend/app/service/postService.ts [hunk 6]
```diff
@@ -188,11 +183,27 @@ class PostService {
...
         else
             return res.status(404).json({ 'message': 'No likes found for that post id' })
     }
 
+    public getRecentUserLikes = async (req: Request, res: Response): Promise<any> => {
+        const foundUser = await this.profileDAO.getProfileByUsername(req.params.username)
+        const foundLikes = await this.dao.getRecentUserLikesByProfileId(foundUser.id, 8)
+
+        if (req.params.username && foundLikes) {
+            const likes = []
```
Comment (AjobK) [inline backend/app/service/postService.ts:193]: newline under

File: frontend/src/components/postLike/index.js [hunk 7]
```diff
@@ -3,19 +3,31 @@ import styles from './postLike.scss'
 import { inject, observer } from 'mobx-react'
 import Axios from 'axios'
 import { withRouter } from 'react-router'
+import { PostLikesList } from '../../components'
+import Like from '../../static/icons/heart-solid.svg';
```
Comment (AjobK) [inline frontend/src/components/postLike/index.js]: Please use icons differently. Use navigation for reference. There is a seperate component for icons

File: frontend/src/components/postLike/index.js [hunk 8]
```diff
@@ -30,31 +42,50 @@ class PostLike extends Component {
...
+                    <p className={`${styles.postLikesAmount}`}>0 likes</p>
+                )}
+                {likesAmount > 0 && (
+                    <p className={`${styles.postLikesAmount} ${styles.clickableLikes}`} onClick={ this.openLikesList }>{likesAmount} {likesAmount === 1 ? 'like' : 'likes'}</p>
+                )}
                 <button onClick={ this.likeClicked }>
-                    &#10084;
+                    <img src={Like} alt={'Like'}/>
                 </button>
+                {this.state.showLikes && (
```
Comment (AjobK) [inline frontend/src/components/postLike/index.js]: Space before 'this.state.showLikes' to make it more readable and clean. Same for cases above such as '{likesAmount' which should be '{ likesAmount'

File: frontend/src/components/postLikesList/index.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            likes: []
+        }
+    }
+
+    componentDidMount() {
+        this.loadLikes()
+    }
+
+    loadLikes = () => {
+        const path = window.location.pathname.split('/').filter(i => i != '').pop()
```
Comment (AjobK) [inline frontend/src/components/postLikesList/index.js:23]: Newline under

File: frontend/src/components/postLikesList/index.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        Axios.get(`${this.props.store.defaultData.backendUrl}/post/like/${path}`)
+            .then(res => {
+                const likes = []
```
Comment (AjobK) [inline frontend/src/components/postLikesList/index.js:27]: Newline under

File: frontend/src/components/postLikesList/index.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                res.data.forEach(like => {
+                    const userLike = {
+                        displayName: like.profile.display_name,
+                        avatarURL: '/src/static/dummy/user/profile.jpg',
+                        level: 0,
+                        title: 'Software Engineer'
+                    }
+                    likes.push(userLike)
```
Comment (AjobK) [inline frontend/src/components/postLikesList/index.js:37]: newline above

File: frontend/src/components/postLikesList/index.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                <div className={`${styles.likesBackground}`} onClick={this.props.closeLikesList}/>
+                <div className={`${styles.likesListWrapper}`}>
+                    <header>
+                        <div className={`${styles.likesListTitle}`}>
+                            <h2>Likes <img src={Like} alt={'Like'} className={`${styles.like}`}/></h2>
+                        </div>
+                        <span className={`${styles.closeLikes}`} onClick={this.props.closeLikesList}>x</span>
+                    </header>
+                    <ul className={`${styles.likesList}`}>
+                        {this.state.likes.map((like, index) => {
```
Comment (AjobK) [inline frontend/src/components/postLikesList/index.js]: space between { and this resulting in '{ this.state.likes.....'

File: frontend/src/pages/profile/index.js [hunk 13]
```diff
@@ -51,6 +55,16 @@ class Profile extends App {
...
   }
 
+  fetchLikedPosts(username) {
+    Axios.get(`${this.props.store.defaultData.backendUrl}/post/liked-by/recent/${username}`)
+        .then((json) => {
+          this.setState({ likes: json.data })
+        })
+        .catch(() => {
+          // this.setState({ error: true })
+        })
```
Comment (AjobK) [inline frontend/src/pages/profile/index.js]: Remove comment

File: frontend/src/components/postLike/index.js (hunk 7 above)
Comment (AjobK) [reply frontend/src/components/postLike/index.js]: Also remove semicolon

File: frontend/src/components/postLikesList/index.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
+import React, { Component } from 'react'
+import styles from './postLikesList.scss'
+import { inject, observer } from 'mobx-react'
+import Axios from "axios"
+import { Link } from "react-router-dom";
```
Comment (AjobK) [inline frontend/src/components/postLikesList/index.js]: Please use single quotes and remove semicolon

File: frontend/src/components/postLikesList/index.js [hunk 15]
```diff
@@ -0,0 +1,78 @@
+import React, { Component } from 'react'
+import styles from './postLikesList.scss'
+import { inject, observer } from 'mobx-react'
+import Axios from "axios"
```
Comment (AjobK) [inline frontend/src/components/postLikesList/index.js]: Single quotes

**Review by AjobK — CHANGES_REQUESTED** [review]
A few other minor things

**Comment by jerohero** [comment]
@AjobK

File: backend/app/service/postService.ts [hunk 16]
```diff
@@ -188,11 +179,29 @@ class PostService {
...
+
+            foundLikes.forEach((like) => {
+                likes.push(like.post)
+            })
+
+            return res.status(200).json(likes)
+        }
```
Comment (AjobK) [inline backend/app/service/postService.ts]: else on this line

File: frontend/src/components/postBanner/index.js [hunk 17]
```diff
@@ -2,7 +2,7 @@ import React, { Component } from 'react'
 import styles from './postBanner.scss'
 import { inject, observer } from 'mobx-react'
 import { Icon } from '..';
```
Comment (AjobK) [inline frontend/src/components/postBanner/index.js:4]: semicolon

**Review by AjobK — CHANGES_REQUESTED** [review]
minor changes and show like for owner of post

**Review by AjobK — APPROVED** [review]
Looks good! :-)

**Review by ryankroon00 — APPROVED** [review]

**Pull request merged**